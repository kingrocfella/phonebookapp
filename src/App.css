.display-end {
  display: flex;
  justify-content: flex-end;
}

.user-avatar {
  background-color: blue;
  border-radius: 50%;
  width: 3rem;
  height: 3rem;
  color: white;
}

.cursor {
  cursor: pointer;
}
