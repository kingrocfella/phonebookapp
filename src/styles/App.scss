@import "./variables";

html,
body {
  background-color: $secondary-color;
}

.display-end {
  display: flex;
  justify-content: flex-end;
}

.display-center-end {
  display: flex;
  justify-content: flex-end;
  align-items: center;
}

.primary-button {
  background-color: $primary-color;
  color: white;
  border-radius: 1.5rem;
  padding: $btnpadding;
}

.block-button {
  background-color: $primary-color;
  color: white;
}

.delete-button {
  background-color: $danger;
  color: white;
  border-radius: 1.5rem;
  padding: $btnpadding;
}

.user-avatar {
  background-color: $primary-color;
  border-radius: 50%;
  width: 3rem;
  height: 3rem;
  color: white;
}

.cursor {
  cursor: pointer;
}

.display-center {
  display: flex;
  justify-content: center;
  align-items: center;
}

.display-vertical-center {
  display: flex;
  align-items: center;
}

.hovered-row {
  background-color: $cardboxshadow;
}

.contact-table {
  border-spacing: 0rem;
  border-collapse: separate;
}

.space-between {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.invisible {
  visibility: hidden;
}
