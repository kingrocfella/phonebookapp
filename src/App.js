import { useState } from "react";
import Modal from "./components/Modal";
import AddContact from "./components/addContact";
import ContactList from "./components/contactList";
import ContactDetails from "./components/contactDetails";
import { ToastContainer } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import "./App.css";

function App() {
  const [contacts, handleContacts] = useState([]);
  const [showModal, handleShowModal] = useState(false);
  const [title, handleTitle] = useState("Contact Information");
  const [modalType, handleModalType] = useState("");
  const [selectedID, setSelectedID] = useState("");
  const [deleteBtn, setDeleteBtn] = useState(false);
  const [detail, setDetail] = useState([]);

  const CONSTANTS = {
    contact: "1",
    viewdetails: "2",
    editdetails: "3",
    deletedetails: "4",
  };

  const addContacts = (data) => {
    handleShowModal(false);
    const arr = [...contacts, data];
    handleContacts(arr);
  };

  const handleViewDetails = (id) => {
    const detail = contacts.filter((contact) => contact.id === id)[0];
    setDetail(detail);
    handleModalType(CONSTANTS.viewdetails);
    handleShowModal(true);
    setSelectedID(id);
  };

  const handleEditContact = () => {
    handleModalType(CONSTANTS.editdetails);
    handleTitle("Edit Contact");
  };

  const saveEditedContact = (data) => {
    const contactList = contacts.filter((contact) => contact.id !== selectedID);
    const newList = [...contactList, data];
    handleContacts(newList);
    handleShowModal(false);
  };

  const handleCheckboxClick = (id) => {
    let arr = [...contacts];
    let checked = false;
    arr.forEach((contact) => {
      if (contact.id === id) contact.check = !contact.check;
      if (contact.check) checked = true;
    });
    handleContacts(arr);
    setDeleteBtn(checked);
  };

  const getCheckedContacts = () => {
    let total = 0;
    contacts.forEach((contact) => {
      if (contact.check) total += 1;
    });
    return total;
  };

  const handleDeleteContact = () => {
    const newarr = contacts.filter((contact) => !contact.check);
    handleContacts(newarr);
    handleShowModal(false);
    setDeleteBtn(false);
  };

  return (
    <>
      <ToastContainer />
      {showModal && (
        <Modal
          show={showModal}
          handleClose={() => handleShowModal(false)}
          title={deleteBtn ? "Delete Contact(s)" : title}
          body={
            modalType === CONSTANTS.contact ||
            modalType === CONSTANTS.editdetails ? (
              <AddContact
                addContacts={addContacts}
                enableEditing={modalType === CONSTANTS.editdetails}
                saveEditedContact={saveEditedContact}
                details={detail}
              />
            ) : modalType === CONSTANTS.viewdetails ? (
              <ContactDetails
                details={detail}
                handleEditContact={handleEditContact}
              />
            ) : modalType === CONSTANTS.deletedetails ? (
              <div className="text-center m-5">
                You are about to delete {getCheckedContacts()} contact(s).
              </div>
            ) : null
          }
          footer={
            deleteBtn ? (
              <button className="btn btn-danger" onClick={handleDeleteContact}>
                CONFIRM
              </button>
            ) : null
          }
        />
      )}
      <div className="container p-5">
        <div className="display-end mt-4">
          <button
            className={`btn ${deleteBtn ? "btn-danger" : "btn-primary"}`}
            onClick={() => {
              handleShowModal(true);
              handleModalType(
                deleteBtn ? CONSTANTS.deletedetails : CONSTANTS.contact
              );
            }}
          >
            {deleteBtn ? "Delete Checked Contact(s)" : "Add Contact"}
          </button>
        </div>
        <div className="mt-5">
          {contacts.length > 0 ? (
            <ContactList
              arr={contacts}
              handleViewDetails={handleViewDetails}
              handleCheckboxClick={handleCheckboxClick}
            />
          ) : (
            <div className="text-center pt-5">You have no contacts!</div>
          )}
        </div>
      </div>
    </>
  );
}

export default App;
