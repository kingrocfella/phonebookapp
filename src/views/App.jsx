import { useState } from "react";
import Modal from "../components/Modal";
import AddContact from "./addContact";
import ContactList from "./contactList";
import ContactDetails from "./contactDetails";
import { ToastContainer, toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import "../styles/App.scss";
import {
  BsPersonPlus,
  BsPersonDash,
  BsExclamationCircle,
} from "react-icons/bs";

function App() {
  const [contacts, handleContacts] = useState([]);
  const [showModal, handleShowModal] = useState(false);
  const [title, handleTitle] = useState("Contact Information");
  const [modalType, handleModalType] = useState("");
  const [selectedID, setSelectedID] = useState("");
  const [deleteBtn, setDeleteBtn] = useState(false);
  const [detail, setDetail] = useState({});

  const CONSTANTS = {
    contact: "1",
    viewdetails: "2",
    editdetails: "3",
    deletedetails: "4",
  };

  const addContacts = (data) => {
    handleShowModal(false);
    const arr = [...contacts, data];
    handleContacts(arr);
    toast.success("Contact created!", {
      position: toast.POSITION.TOP_RIGHT,
    });
  };

  const handleViewDetails = (id) => {
    const detail = contacts.filter((contact) => contact.id === id)[0];
    setDetail(detail);
    handleModalType(CONSTANTS.viewdetails);
    handleShowModal(true);
    setSelectedID(id);
  };

  const handleEditContact = () => {
    handleModalType(CONSTANTS.editdetails);
    handleTitle("Edit Contact");
  };

  const saveEditedContact = (data) => {
    const contactList = contacts.filter((contact) => contact.id !== selectedID);
    const newList = [...contactList, data];
    handleContacts(newList);
    handleShowModal(false);
    toast.success("Contact successfully edited!", {
      position: toast.POSITION.TOP_RIGHT,
    });
  };

  const handleCheckboxClick = (id) => {
    let arr = [...contacts];
    let checked = false;
    arr.forEach((contact) => {
      if (contact.id === id) contact.check = !contact.check;
      if (contact.check) checked = true;
    });
    handleContacts(arr);
    setDeleteBtn(checked);
  };

  const getCheckedContacts = () => {
    let total = 0;
    contacts.forEach((contact) => {
      if (contact.check) total += 1;
    });
    return total;
  };

  const handleDeleteContact = () => {
    const newarr = contacts.filter((contact) => !contact.check);
    handleContacts(newarr);
    handleShowModal(false);
    setDeleteBtn(false);
    toast.success("Contact(s) deleted!", {
      position: toast.POSITION.TOP_RIGHT,
    });
  };

  return (
    <>
      <ToastContainer />
      {showModal && (
        <Modal
          show={showModal}
          handleClose={() => handleShowModal(false)}
          title={
            deleteBtn
              ? "Delete Contact(s)"
              : modalType === CONSTANTS.contact
              ? "Add Contact"
              : title
          }
          body={
            modalType === CONSTANTS.contact ||
            modalType === CONSTANTS.editdetails ? (
              <AddContact
                addContacts={addContacts}
                enableEditing={modalType === CONSTANTS.editdetails}
                saveEditedContact={saveEditedContact}
                details={detail}
              />
            ) : modalType === CONSTANTS.viewdetails ? (
              <ContactDetails
                details={detail}
                handleEditContact={handleEditContact}
              />
            ) : modalType === CONSTANTS.deletedetails ? (
              <div className="text-center m-5">
                You are about to delete {getCheckedContacts()} contact(s).
              </div>
            ) : null
          }
          footer={
            deleteBtn ? (
              <button
                className="btn delete-button"
                onClick={handleDeleteContact}
              >
                CONFIRM
              </button>
            ) : null
          }
        />
      )}
      <div className="container text-white">
        <div className="space-between mt-5">
          <h3>Phonebook App</h3>
          <button
            className={`btn datatest-2 ${deleteBtn ? "delete-button" : "primary-button"}`}
            onClick={() => {
              handleShowModal(true);
              handleModalType(
                deleteBtn ? CONSTANTS.deletedetails : CONSTANTS.contact
              );
            }}
          >
            {deleteBtn ? <BsPersonDash /> : <BsPersonPlus />} &nbsp;&nbsp;
            {deleteBtn ? "Delete Checked Contact(s)" : "Add Contact"}
          </button>
        </div>
        <div className="mt-5">
          {contacts.length > 0 ? (
            <ContactList
              arr={contacts}
              handleViewDetails={handleViewDetails}
              handleCheckboxClick={handleCheckboxClick}
            />
          ) : (
            <div className="text-center pt-5 datatest-1">
              <BsExclamationCircle /> &nbsp; You have no contacts!
            </div>
          )}
        </div>
      </div>
    </>
  );
}

export default App;
