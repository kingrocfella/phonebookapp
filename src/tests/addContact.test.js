import React from "react";
import { shallow, mount } from "enzyme";
import App from "../views/addContact";
import { findClassAtrr } from "./utils";

const mockCallBack = jest.fn();
const props = {
  details: {
    firstName: "test",
    lastName: "test",
    phoneNumber: "test",
  },
  addContacts: mockCallBack,
  enableEditing: false,
  saveEditedContact: mockCallBack,
};
const container = shallow(<App {...props} />);
const mountContainer = mount(<App {...props} />);

describe("<App />", () => {
  it("should match the snapshot", () => {
    expect(container.html()).toMatchSnapshot();
  });

  it("Should have proper props for text and tel fields", () => {
    expect(container.find('input[type="text"]').length).toEqual(2);
    expect(findClassAtrr(container, "datatest-4").length).toEqual(1);
  });

  it("It should have a submit button", () => {
    expect(container.find('button[type="submit"]').length).toEqual(1);
  });

  it("Submit button should have test of SUBMIT and be disabled", () => {
    const btn = findClassAtrr(mountContainer, "datatest-3")
    expect(btn.get(0).props.children).toBe("SUBMIT");
    expect(btn.get(0).props.disabled).toBe(true);
  });
});
