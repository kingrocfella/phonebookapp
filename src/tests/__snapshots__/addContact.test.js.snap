// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match the snapshot 1`] = `"<div class=\\"container\\"><form><div class=\\"form-row\\"><div class=\\"form-group col-md-6\\"><label for=\\"firstname\\">First Name</label><input type=\\"text\\" value=\\"\\" class=\\"form-control datatest-5\\" placeholder=\\"First Name\\" aria-describedby=\\"first name\\"/></div><div class=\\"form-group col-md-6\\"><label for=\\"lastname\\">Last Name</label><input type=\\"text\\" value=\\"\\" class=\\"form-control datatest-6\\" placeholder=\\"Last Name\\" aria-describedby=\\"last name\\"/></div></div><div class=\\"form-group datatest-4\\"><label for=\\"phonenumber\\">Phone Number</label><div class=\\" react-tel-input\\"><div class=\\"special-label\\">Phone</div><input type=\\"tel\\" class=\\"form-control\\" style=\\"width:100%\\" value=\\"\\" placeholder=\\"1 (702) 123-4567\\"/><div class=\\" flag-dropdown\\"><div class=\\"selected-flag\\" title=\\"\\" tabindex=\\"0\\" role=\\"button\\" aria-haspopup=\\"listbox\\"><div class=\\"flag 0\\"><div class=\\"arrow\\"></div></div></div></div></div></div><button type=\\"submit\\" class=\\"mt-5 btn block-button btn-block datatest-3\\" disabled=\\"\\">SUBMIT</button></form></div>"`;
